File: code/scenes/src/main/java/org/fbme/scenes/cells/EditorCell_SceneLabel.java
package org.fbme.scenes.cells;

import jetbrains.mps.editor.runtime.style.StyleAttributes;
import jetbrains.mps.nodeEditor.cells.EditorCell_Basic;
import jetbrains.mps.nodeEditor.cells.ParentSettings;
import jetbrains.mps.nodeEditor.cells.TextLine;
import jetbrains.mps.openapi.editor.EditorContext;
import org.jetbrains.mps.openapi.model.SNode;

import java.awt.*;

public class EditorCell_SceneLabel extends EditorCell_Basic {

    private final TextLine myTextLine;

    public EditorCell_SceneLabel(EditorContext context, SNode node, String text, boolean error) {
        super(context, node);
        myTextLine = new TextLine(text, getStyle(), false, context);
        if (error) {
            myTextLine.showErrorColor();
        }
    }

    @Override
    protected void relayoutImpl() {
        myTextLine.relayout();
        myWidth = myTextLine.getWidth();
        myHeight = myTextLine.getHeight();
    }

    @Override
    public void setSelected(boolean selected) {
        super.setSelected(selected);
        getStyle().set(StyleAttributes.UNDERLINED, selected);
    }

    @Override
    public int getDescent() {
        return myTextLine.getDescent();
    }

    @Override
    protected void paintContent(Graphics graphics, ParentSettings settings) {
        myTextLine.paint(graphics, myX, myY);
    }

    @Override
    public void paintSelection(Graphics g, Color c, boolean drawBorder, ParentSettings parentSettings) {
        // do noting
    }
}

File: code/scenes/src/main/kotlin/org/fbme/scenes/cells/EditorCell_SceneLabel.kt
package org.fbme.scenes.cells

import jetbrains.mps.editor.runtime.style.StyleAttributes
import jetbrains.mps.nodeEditor.cells.EditorCell_Basic
import jetbrains.mps.nodeEditor.cells.ParentSettings
import jetbrains.mps.nodeEditor.cells.TextLine
import jetbrains.mps.openapi.editor.EditorContext
import org.jetbrains.mps.openapi.model.SNode
import java.awt.Color
import java.awt.Graphics

class EditorCell_SceneLabel(
    context: EditorContext,
    node: SNode?,
    text: String?,
    error: Boolean
) : EditorCell_Basic(context, node) {
    private val textLine: TextLine
    override fun relayoutImpl() {
        textLine.relayout()
        myWidth = textLine.width
        myHeight = textLine.height
    }

    override fun setSelected(selected: Boolean) {
        super.setSelected(selected)
        style.set(StyleAttributes.UNDERLINED, selected)
    }

    override fun getDescent(): Int {
        return textLine.descent
    }

    override fun paintContent(graphics: Graphics, settings: ParentSettings) {
        textLine.paint(graphics, myX, myY)
    }

    override fun paintSelection(g: Graphics, c: Color, drawBorder: Boolean, parentSettings: ParentSettings) {
        // do noting
    }

    init {
        textLine = TextLine(text, style, false, editor.editorComponentSettings)
        if (error) {
            textLine.showErrorColor()
        }
    }
}
